### PR TITLE
PyGRB: Single IFO running

### PR DESCRIPTION
**Pull request by a-r-williamson**
These commits will allow the pygrb workflow generator to fall back to a single IFO search if there isn't enough data for a coherent search. All that is required for this to happen is to include the option 'allow-single-ifo-search' in the [workflow] section of the configuration file. If this isn't given, the code will exit in the case of insufficient coherent data, which was the previous behaviour.

Since this functionality was designed with the online search in mind, where a single set of static config files are desired, the LegacyCohPTFInspiralExecutable class will check for any options that have been passed to it that aren't compatible with single IFO running. These include face-on/-away options and short timeslide options. It will also add the --do-sngl-chi-tests argument if there is only 1 IFO.

### CODE REVIEW

**Comment by spxiwh** [comment]
Hi Andrew,

This only affects the GRB code so you are free to merge this, and the changes to fall back to single detector running in the segment stuff all looks good.

..... *BUT*, one of the main principles when writing and moving to the PyCBC workflow interface was that options are not, and should not, be hidden. In some cases (ie. file inputs) this is difficult to avoid, but I don't think it is necessary to hide things like --do-sngl-chi-tests and the face-on/face-away options.

My suggestion would be to use the tagging system. Create a:

[cohptfinspiral-sngl]

and

[cohptfinspiral-coherent]

sections and use these to supply options that would only be given in the coherent, or single detector, cases. Then the workflow generator can add the appropriate tag when calling the matched-filter module depending on how many ifos are in play.

**Comment by a-r-williamson** [comment]
Hi Ian,

Thanks, that's a much nicer solution, I agree. I'll have a go at changing this.

**Comment by a-r-williamson** [comment]
Hi Ian,

Let me know if this looks better. The options are now split into separate config file sections so none are added or removed in legacy_ihope.py

Andrew

File: bin/pygrb/pycbc_make_offline_grb_workflow [hunk 1]
```diff
@@ -173,7 +194,7 @@ if wflow.cp.has_section("workflow-injections"):
     # Generate injection matched filter workflow
     inj_insp_files = _workflow.setup_matchedfltr_workflow(wflow, sciSegs,
             datafind_veto_files, inj_splitbank_files, injDir, injs,
-            tags=["injections"])
+            tags=[mf_tag + "_injections"])
```
Comment (spxiwh) [inline bin/pygrb/pycbc_make_offline_grb_workflow]: I think these should be two tags. Ie 
```
tags=[mf_tag,"injections"]
```

This then let's you specify options specific for each tag (ie. the injection-specfic options don't care on the mf_tag and vice-versa).

File: bin/pygrb/pycbc_make_offline_grb_workflow [hunk 2]
```diff
@@ -210,9 +231,8 @@ if wflow.cp.has_section("workflow-injections"):
 # MAIN MATCHED FILTERING
 inspDir = os.path.join(currDir, "inspiral")
 inspiral_files = _workflow.setup_matchedfltr_workflow(wflow, sciSegs,
-                                                      datafind_veto_files,
-                                                      splitbank_files, inspDir,
-                                                      tags=["no_injections"])
+        datafind_veto_files, splitbank_files, inspDir,
+        tags=[mf_tag + "_no_injections"])
```
Comment (spxiwh) [inline bin/pygrb/pycbc_make_offline_grb_workflow]: Same comment here:

```
tags=[mf_tag,"no_injections"]
```

File: examples/workflow/pygrb/ER8/er8_main_offline.ini [hunk 3]
```diff
@@ -113,25 +111,30 @@ num-auto-chisq-points = 40
...
 do-clustering =
 cluster-window = 0.1
 inverse-spec-length = 32
-face-on-analysis =
-face-away-analysis =
 
-[inspiral-no_injections]
-;do-short-slides =
-;short-slide-offset = 6
+[inpiral-sngl_no_injections&inspiral-sngl_injections]
```
Comment (spxiwh) [inline examples/workflow/pygrb/ER8/er8_main_offline.ini]: This then becomes [inspiral-sngl]

File: examples/workflow/pygrb/ER8/er8_main_offline.ini [hunk 4]
```diff
@@ -113,25 +111,30 @@ num-auto-chisq-points = 40
...
+do-sngl-chi-tests =
+
+[inspiral-coherent_no_injections]
+do-short-slides =
+short-slide-offset = 6
 
-[inspiral-injections]
+[inspiral-sngl_injections&inspiral-coherent_injections]
```
Comment (spxiwh) [inline examples/workflow/pygrb/ER8/er8_main_offline.ini]: and this remains [inspiral-injections]

File: examples/workflow/pygrb/ER8/er8_main_offline.ini [hunk 5]
```diff
@@ -113,25 +111,30 @@ num-auto-chisq-points = 40
...
 
-[inspiral-no_injections]
-;do-short-slides =
-;short-slide-offset = 6
+[inpiral-sngl_no_injections&inspiral-sngl_injections]
+do-sngl-chi-tests =
+
+[inspiral-coherent_no_injections]
+do-short-slides =
+short-slide-offset = 6
```
Comment (spxiwh) [inline examples/workflow/pygrb/ER8/er8_main_offline.ini]: Ah balls, this option is needed only for the coherent *and* no injections case. Okay, maybe the way you do it is better for now until we can find a nice way of specifying that an option should only be specified if *two* tags are present.

File: pycbc/workflow/legacy_ihope.py [hunk 6]
```diff
@@ -462,7 +471,6 @@ def __init__(self, cp, name, universe=None, ifo=None, injection_file=None,
                  out_dir=None, tags=[]):
         super(LegacyCohPTFInjcombiner, self).__init__(cp, name, universe,
               ifo=ifo, out_dir=out_dir, tags=tags)
-        self.cp = cp
         self.ifos = ifo
```
Comment (spxiwh) [inline pycbc/workflow/legacy_ihope.py]: As you are removing things, note that:

self.ifo_list (list of ifos (['H1','L1']))
self.ifo_string ('H1L1')

and

self.ifo (self.ifo_list if len(self.ifo_list) == 1 else raise ValueError)

are all set by the parent class, and self.ifos can be removed and replaced by one of these where needed.

**Comment by spxiwh** [comment]
Hi Andrew,

Thanks for the updated patch. This is better. It really seems like injections/no_injections and sngl/coherent should be separate tags, but I do see that one argument is given only if injections *and* coherent tags are used. Currently we have no nice way of specifying this in a configuration file. If you have an idea for that, it could be added to:

https://github.com/ligo-cbc/pycbc/blob/master/pycbc/workflow/configuration.py#L547

but if you want to look at that (or get one of the other Cardiff folk to look at it) ping @ahnitz first in case he has any thoughts.

I also noted a place where you set self.ifos, which is a duplicate of the parent class (you already removed some other cases where you are doing that!)

I would suggest that you merge this now (you have permission on github to merge, right?) and then look to addressing these comments in future changes.

Cheers
Ian

**Comment by a-r-williamson** [comment]
Thanks Ian. I initially did this with the two tags, as that would be far nicer, however I realised that the slides would be the issue as you also note. I agree it would be really nice to get a "only if all these tags are present" sort of syntax so I'll look into that if I get the chance.

The self.ifos thing is noted. Will address this soon also.

Cheers,
Andrew

**Pull request merged**